### PR TITLE
Reference implementation of defer and stream spec

### PR DESCRIPTION
**Pull request by robrichard**
Continued from https://github.com/graphql/graphql-js/pull/2839

This is the reference impementation of the defer/stream spec proposal. The corresponding PR for spec changes are here: https://github.com/graphql/graphql-spec/pull/742

Please limit comments on this PR to code review of this implementation. Discussion on the defer/stream spec is located in a dedicated repo: https://github.com/robrichard/defer-stream-wg/discussions

### CODE REVIEW

**Comment by robrichard** [comment]
@github-actions publish-pr-on-npm

**Comment by glasser** [comment]
@github-actions publish-pr-on-npm

**Comment by glasser** [comment]
@robrichard @IvanGoncharov I'm finding the types a bit hard to follow here.

So `execute` can *either* return (perhaps via Promise) a single ExecutionResult or an async generator of AsyncExecutionResults. That might make me think that the former case happens when incremental delivery is not involved, and the latter happens with incremental delivery.

However, ExecutionResult itself has `hasNext` and `incremental` so I guess that's not true. But maybe if you get an ExecutionResult directly from `execute` it won't have those fields set?

AsyncExecutionResult can be ExecutionResult or SubsequentExecutionResult. The difference is that only ExecutionResult has `errors` and `data`; SubsequentExecutionResult has those nested inside `incremental`.

But I don't get why `ExecutionResult` can have `errors` and `data` *both* at the top level *and* nested inside `incremental`.

I also don't get why `ExecutionResult` *and* `SubsequentExecutionResult` have `extension` both at the top level and nested inside `incremental`.

This would make a lot more sense to me if there was the "non incremental delivery" return value that only has data/errors/extensions at the top level, and the "incremental delivery" return value that only has data/errors/extensions nested inside "incremental". Am I missing something?

I'll take a look at the source too to see if that helps me understand it better.

File: src/execution/execute.ts [hunk 1]
```diff
@@ -1294,3 +1553,444 @@ function assertEventStream(result: unknown): AsyncIterable<unknown> {
...
+      // a different call to next has exhausted all payloads
+      return { value: undefined, done: true };
+    }
+    const index = exeContext.subsequentPayloads.indexOf(asyncPayloadRecord);
+    if (index === -1) {
+      // a different call to next has consumed this payload
+      return race();
+    }
+
+    exeContext.subsequentPayloads.splice(index, 1);
```
Comment (glasser) [inline src/execution/execute.ts:1834]: How about using a `Set` instead of an array for subsequentPayloads, so that this call (and the indexOf above) work in constant time? Otherwise running n defers may take O(n^2) time if they end in pessimal order.

File: src/execution/execute.ts [hunk 2]
```diff
@@ -1294,3 +1553,444 @@ function assertEventStream(result: unknown): AsyncIterable<unknown> {
...
+): AsyncGenerator<AsyncExecutionResult, void, void> {
+  let _hasReturnedInitialResult = false;
+  let isDone = false;
+
+  async function race(): Promise<IteratorResult<AsyncExecutionResult>> {
+    if (exeContext.subsequentPayloads.length === 0) {
+      // async iterable resolver just finished and no more pending payloads
+      return {
+        value: {
+          hasNext: false,
```
Comment (glasser) [inline src/execution/execute.ts:1802]: Why is this different from the length 0 case directly in `next()`, which returns `{ value: undefined, done: true }`?

File: src/execution/flattenAsyncIterable.ts [hunk 3]
```diff
@@ -0,0 +1,50 @@
...
+    }
+    const result = await currentIterator.next();
+    if (result.done) {
+      iteratorStack.shift();
+      return next();
+    } else if (isAsyncIterable(result.value)) {
+      const childIterator = result.value[
+        Symbol.asyncIterator
+      ]() as AsyncIterator<T>;
+      iteratorStack.unshift(childIterator);
```
Comment (glasser) [inline src/execution/flattenAsyncIterable.ts:31]: My assumption is that `unshift` is more likely to be `O(n)` than `push`. (I didn't bother to write my own benchmark but [this one I quickly found](https://www.measurethat.net/Benchmarks/Show/16089/0/pop-vs-shift-vs-unshift-vs-pop) certainly makes (in Chrome v8) push look much better than unshift.) Consider switching to push/pop.

File: src/execution/flattenAsyncIterable.ts [hunk 4]
```diff
@@ -0,0 +1,50 @@
...
+      return next();
+    }
+    return result;
+  }
+  return {
+    next,
+    return() {
+      iteratorStack = [];
```
Comment (glasser) [inline src/execution/flattenAsyncIterable.ts:39]: should this `return` the iterators in the stack first? (ditto for `throw` below.)

**Comment by glasser** [comment]
I tried to simplify the types as described above. https://github.com/graphql/graphql-js/pull/3703 is what I came up with.

A lot of the complexity comes from the `incremental` field; this is a pretty recent addition, right? I'm not sure I really understand why it exists rather than the fields on it going directly on multiple top-level messages. Something about being able to deliver multiple deferred chunks at once? Does `incremental` get actually written in the JSON on the wire?

File: src/execution/flattenAsyncIterable.ts (hunk 3 above)
Comment (glasser) [reply src/execution/flattenAsyncIterable.ts:31]: (although note that in #3703 I refactored things so that this function only needs to flatten precisely 2-level iterators with a single type that is always returned)

File: src/execution/execute.ts [hunk 5]
```diff
@@ -139,9 +150,45 @@ export interface FormattedExecutionResult<
 > {
   errors?: ReadonlyArray<GraphQLFormattedError>;
   data?: TData | null;
+  hasNext?: boolean;
+  extensions?: TExtensions;
+  incremental?: ReadonlyArray<IncrementalResult>;
```
Comment (glasser) [inline src/execution/execute.ts:155]: `IncrementResult.errors` is a `ReadonlyArray<GraphQLError>` (not `ReadonlyArray<GraphQLFormattedError>`) so that's not a good fit for `FormattedExecutionResult`. I do find FormattedExecutionResult to be a useful type so that suggests we need Formatted versions of the Incremental types.

**Comment by glasser** [comment]
Having tried to implement consuming in Apollo Server, I'm increasingly convinced that (assuming we don't want the first message to have an `incremental` field, which [seems to be the consensus](https://github.com/robrichard/defer-stream-wg/discussions/41#discussioncomment-2915426)) we should make `execute` (or its replacement) always return a "first execution result" (which should never have an `incremental` field), and then if anything is deferred/streamed, also return an async iterator of "subsequent execution results" (which should always have an `incremental` field). That works a lot better from the consumer side than the current async iterator which is "never empty and the first one has a different type from the rest". I would also much rather have a nice strongly typed optional field with the subsequent result iterator than have to probe a result to figure out what kind it is (basically requiring me to copy and paste isAsyncIterator into my codebase).

File: src/execution/execute.ts (hunk 2 above)
Comment (glasser) [reply src/execution/execute.ts:1802]: Oh, I get it. This pushed an actual `{hasNext: false}` message to clients; then the next one closes the iterator.

File: src/execution/flattenAsyncIterable.ts [hunk 6]
```diff
@@ -0,0 +1,50 @@
...
+  const iteratorMethod = iterable[Symbol.asyncIterator];
+  const iterator: any = iteratorMethod.call(iterable);
+  let iteratorStack: Array<AsyncIterator<T>> = [iterator];
+
+  async function next(): Promise<IteratorResult<T | AT, void>> {
+    const currentIterator = iteratorStack[0];
+    if (!currentIterator) {
+      return { value: undefined, done: true };
+    }
+    const result = await currentIterator.next();
```
Comment (glasser) [inline src/execution/flattenAsyncIterable.ts:23]: flattenAsyncIterable is buggy if you call `next` multiple times in parallel.
https://github.com/graphql/graphql-js/pull/3709

Comment (glasser) [reply src/execution/flattenAsyncIterable.ts]: @robrichard It looks like you rewrote flattenAsyncIterable as a simple async generator. Note that that has the problem that you can't stop (`return`) the generator while it's blocked on reading (ie anywhere other than at the `yield`. That was causing issues in one of the subscribe tests for me when I tried it that way, and in fact I see that it's failing for you too (ie CI is failing). I think this is a fundamental issue with using async generator functions that wrap other async iterators (see eg https://github.com/tc39/proposal-async-iteration/issues/126)

In https://github.com/graphql/graphql-js/pull/3703 I do a few things — main motivation is simplifying/explicit-ifying the types used, but I also think I fix flattenAsyncIterable without introducing the `return` problem. Though I'll be honest, all that stuff around stopping generators is weird and I can believe I haven't implemented it quite correctly either.

**Comment by yaacovCR** [comment]
We've hit a few concurrency bugs already (#2975, #3710) with the "raw" async generator objects currently used within incremental delivery.

In #3694, we have a PR to introduce @brainkim 's [repeaters](https://repeater.js.org/) into `graphql-js` to ensure that our returned async generator objects actually behave exactly like async generators by design. In particular:

In particular:
1. Calls to `next()`, `return()`, and `throw()` will not interfere with each other; they will be executed sequentially.
2. The promises returned by `next()`, `return()`, and `throw()` will all settle in call order.

[`graphql-executor`](https://github.com/yaacovCR/graphql-executor/) builds on Repeaters to create a [publisher/dispatcher abstraction](https://github.com/yaacovCR/graphql-executor/blob/main/src/execution/publisher.ts) that allows the executor to push results to the generator as they are ready, without the need for any complicated promise racing.

#3694 just uses Repeaters to implement `mapAsyncIterable`, but graphql-executor does the same for `flattenAsyncIterable`, (as well as using them for the publisher/dispatcher).

File: src/execution/flattenAsyncIterable.ts (hunk 3 above)
Comment (glasser) [reply src/execution/flattenAsyncIterable.ts]: (this is now gone)

File: src/execution/execute.ts (hunk 1 above)
Comment (glasser) [reply src/execution/execute.ts:2054]: https://github.com/graphql/graphql-js/pull/3717

**Comment by robrichard** [comment]
@github-actions publish-pr-on-npm

**Comment by glasser** [comment]
One option is to just avoid async iterators entirely. @benjamn privately suggested to me the idea that each chunk just contains a `next(): Promise<Chunk>`; there's no complexity about parallel calls here because each `next()` call on the same chunk always would return the same next chunk. You'd need some other explicit `close()` or `cancel()` or whatever call. The main advantage of async iterators to me is just that it's how graphql-js already works for subscriptions.

Fortunately this is purely a question about how to handle this in graphql-js, not something that should block the overall of the spec update from merging. I think sticking with the same API as subscriptions makes sense, and graphql-js could consider later creating a second new API (both could be supported) for following these streams if we want.

File: src/execution/execute.ts [hunk 7]
```diff
@@ -1294,3 +1771,461 @@ function assertEventStream(result: unknown): AsyncIterable<unknown> {
...
+      return { value: undefined, done: true };
+    }
+
+    const incremental = getCompletedIncrementalResults(exeContext);
+    const hasNext = exeContext.subsequentPayloads.size > 0;
+
+    if (!incremental.length && hasNext) {
+      return next();
+    }
+    if (!incremental.length && !hasNext) {
```
Comment (glasser) [inline src/execution/execute.ts:2071]: you might find it slightly simpler to replace from here to the end of the function with

```ts
    if (!hasNext) {
      isDone = true;
    }

    return {
      value: incremental.length ? { incremental, hasNext } : { hasNext },
      done: false,
    };
```

**Comment by glasser** [comment]
(BTW my basic opinion about concurrent next/returns is that if we claim to implement AsyncIterator then we should do so correctly, but if we eventually decide a simpler API is good enough for our use case then that's good too. For the specific case of defer/stream results, the ordering between chunks is quite important — the chunks are designed to be applied in order. So the ability to wait on multiple `next` calls in parallel (or to use `return` for anything other than cleanup) is not really important if you ask me.)

**Comment by michaelstaib** [comment]
@robrichard is the new `items` name for the result already implemented? Went through the tests but did not find any result with the `items` field for stream results.

**Comment by robrichard** [comment]
@michaelstaib yes `items` is implemented. See the tests in `stream-test.ts`: https://github.com/graphql/graphql-js/blob/defer-stream/src/execution/__tests__/stream-test.ts#L140

**Comment by michaelstaib** [comment]
Ah, overlooked this one. Thanks for pointing this one out.

File: src/execution/flattenAsyncIterable.ts (hunk 6 above)
Comment (robrichard) [reply src/execution/flattenAsyncIterable.ts]: this was fixed in https://github.com/graphql/graphql-js/pull/3710

**Comment by calvincestari** [comment]
@robrichard - I know this was a while ago but is it correct that `GraphQLDeferDirective` and `GraphQLStreamDirective` were not added to the `specifiedDirectives` list? For schemas that want to use those directives should they manually add them to the config/schema directive lists?
